components/omega/doc/design/Driver.md

```diff
@@ -0,0 +1,253 @@
+   other args as needed (eg forcing)
+)
+```
+The model state, current time and the time interval will be input.
+Other variables will be needed as well, like the surface forcing
+fields, and will be added as needed. On return, the time instant
+will contain the end time of the interval and the end alarm will
+be ringing if the end of the simulation has been reached. The
+CurrState will be the ocean state at the end of the run interval.
+An integer error code will be zero if successful and non-zero if an
+error was encountered. This interface will be modified as needed to
+include other fields.
+
+#### 4.2.3 Finalize
+
+The finalize method will write a checkpoint/restart file
+(if not already written by driver or run method on the
+final timestep) and then clean up all arrays and classes by
+calling the relevant routines for all Omega modules. The
+interface is similar to the Init interface:
+
+```c++
+int OMEGA::OcnFinalize(
+   OMEGA::TimeInstant &CurrTime, ///< [in] current sim time
+   OMEGA::State &CurrState, ///< [in] current model state
+   other args as needed for restart
+);
+```
+An integer return value will be zero if successful and
+non-zero if an error is encountered either writing a
+restart or deallocating memory.
+
+#### 4.2.4 Standalone driver (main)
+
+With the interfaces above, the standalone driver should look
+something like the code below (details subject to change during implementation).
+```c++
+
+int main(int argc, char **argv) {
+
+   MPI_Init(); // initialize MPI
+   yakl::init(); // initialize YAKL
+
+   OMEGA::State CurrState;
+   OMEGA::TimeInstant CurrTime;
+   OMEGA::TimeInterval RunInterval;
+   OMEGA::Alarm EndAlarm;
+
+   int Err = OcnInit(MPI_COMM_WORLD, CurrTime, RunInterval, CurrState,
+                     etc);
+   if (Err != 0) LOG_ERROR("Error initializing OMEGA");
+
+
+   while (Err == 0 && !(EndAlarm.isRinging()) ) {
+
+      // call routines for forcing and other inputs
+
+      // call run method
+      Err = OMEGA::OcnRun(CurrTime, RunInterval, EndAlarm,
+                          CurrState, etc);
+      if (Err != 0) LOG_ERROR("Error advancing Omega one interval");
+
+      // Other tasks if needed (eg IO could occur here or within run
+      // method
+   }
+
+   int Err2 = OMEGA::OcnFinalize(CurrTime, CurrState, etc);
+   if (Err2 != 0) LOG_ERROR("Error finalizing Omega");
+
+   int ErrAll = abs(Err) + abs(Err2);
+   if (ErrAll == 0){
+      LOG_INFO("OMEGA successfully completed");
+   } else {
+      LOG_ERROR("OMEGA terminating due to error");
+   }
+
+   // Exit various environments
+   yakl::finalize();
+   MPI::Finalize();
+
+   return ErrAll;
+}
+
+```
+
+#### 4.2.5 Coupler-component interfaces
+
+To be added later
+
+## 5 Verification and Testing
+
+### 5.1 Test forward model
+
+A forward model test is included as part of the test suite
+that runs the standalone model and tests for successful
+completion. All forward model system testing will be
+inherently testing the driver.
```

Comment:
Do you anticipate this will amount to running a couple timesteps of a very lightweight global case (like a standalone smoke test)? Would we just be testing for a successful error code or doing some kind of a baseline comparison, etc.?

Reply:
Yes, that's correct. Just a simple smoke test to verify the standalone model runs. The current ctest script actually runs a dummy driver already and just checks for completion. We'll rely more on Polaris testing for baseline comparisons. I have clarified this.